### ISSUE TITLE
Doubled horizontal mouse sensitivity in some games

### ISSUE THREAD

**Issue by johnnovak** [dosbox-staging/dosbox-staging#1811]
The horizontal mouse sensitivity is doubled in some games. This is not the end of the world as one can you halve the horizontal sensitivity in the config, but it would be nice to get it fixed if possible. Note that it's unclear whether the issue is present in these games on real hardware or not; that's something to look into as well.

The behaviour is the same in all DOSBox variants I tried (0.74, SVN, DOSBox-X, Staging latest main). A non-exhaustive list of games that are affected:

- Eric the Unready
- Shannara
- Discworld
- Knights of Xentar

As one can kind of adapt to the sensitivity issue after a while, here's a 100% foolproof way to spot it:

1. Use windowed mode _without_ seamless mouse support (this is important; in seamless mode you can't reproduce the problem!)
2. Lock the mouse, and move it diagonally to "paint" a 45 degree line with the pointer.
3. Keep moving the mouse back and forth along the 45 degree line, then press the middle button to unlock it while you keep moving it.
4. You'll notice that using the same motion the 45 degree line will turn into a 60 degree one (so it will be vertically stretched).
 
The problem is present in fullscreen mode too, it's just easier to spot the difference between the wrong and correct behaviour in windowed mode using the above method. None of @FeralChild64's newly introduced mouse features made any difference.

For reference, Eye of the Beholder I and II has perfect and very smooth mouse movement behaviour, so it might be worthwhile to check those games for regressions when making any changes to the mouse code.

**Comment by kcgen** [comment]
I recently hit this in Synergist, a mid 90s point and click adventure. I found I had to set historical sensitivity to rough 65% to get circular results. 

Most of us are using widescreen displays - maybe that wider horizontal ratio at the host level is being "scaled" into the cursor position by sdl? (Hmm... Or maybe not, if this were the case,  everything would be affected the same).

**Comment by johnnovak** [comment]
> Most of us are using widescreen displays - maybe that wider horizontal ratio at the host level is being "scaled" into the cursor position by sdl? (Hmm... Or maybe not, if this were the case, everything would be affected the same).

I don't think that has anything to do with it as most games are perfectly fine. E.g. try with Eye of the Beholder I & II — you'll get perfect and smooth mouse movement with the defaults.

**Comment by FeralChild64** [comment]
@johnnovak , @kcgen I think my development code fixes the problem. Could you check on branch https://github.com/FeralChild64/dosbox-staging/tree/fc/mouse-updates-4 ?

**Comment by kcgen** [comment]
This is looking really nice, @FeralChild64.

Hitting a build-time issue (probably minor):

```
FAILED: src/ints/libints.a.p/mouse_interfaces.cpp.o 
clang++ -fcolor-diagnostics -Werror=format-security -Wextra -Wshadow -Wcast-align -Wunused -Woverloaded-virtual -Wpedantic -Wconversion -Wsign-conversion -Wdouble-promotion -Wformat=2 -fstrict-aliasing -Wstrict-aliasing=2 -Wno-error=newline-eof -Wno-error=incompatible-pointer-types -Wno-error -Weffc++ -Wextra-semi -Wnon-virtual-dtor -Woverloaded-virtual -Isrc/ints/libints.a.p -Isrc/ints -I../../src/ints -I../../include -I. -I../.. -Isrc/libs/ghc -I../../src/libs/ghc -Isrc/libs/loguru -I../../src/libs/loguru -I/usr/include/SDL2 -fcolor-diagnostics -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -Wextra -Wpedantic -std=c++17 -O0 -g -Wno-unknown-pragmas -Weffc++ -Wextra-semi -D_REENTRANT -MD -MQ src/ints/libints.a.p/mouse_interfaces.cpp.o -MF src/ints/libints.a.p/mouse_interfaces.cpp.o.d -o src/ints/libints.a.p/mouse_interfaces.cpp.o -c ../../src/ints/mouse_interfaces.cpp
../../src/ints/mouse_interfaces.cpp:114:2: warning: extra ';' outside of a function is incompatible with C++98 [-Wc++98-compat-extra-semi]
};
 ^
../../src/ints/mouse_interfaces.cpp:120:2: warning: extra ';' outside of a function is incompatible with C++98 [-Wc++98-compat-extra-semi]
};
 ^
../../src/ints/mouse_interfaces.cpp:432:25: error: constexpr variable 'limit_lo' must be initialized by a constant expression
        constexpr float limit_lo = std::log10(1.0f / multiplicity);
                        ^          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/cmath:358:12: note: subexpression not valid in a constant expression
  { return __builtin_log10f(__x); }
           ^
../../src/ints/mouse_interfaces.cpp:432:36: note: in call to 'log10(3.125000e-02)'
        constexpr float limit_lo = std::log10(1.0f / multiplicity);
                                   ^
../../src/ints/mouse_interfaces.cpp:433:25: error: constexpr variable 'limit_hi' must be initialized by a constant expression
        constexpr float limit_hi = std::log10(multiplicity);
                        ^          ~~~~~~~~~~~~~~~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/cmath:358:12: note: subexpression not valid in a constant expression
  { return __builtin_log10f(__x); }
           ^
../../src/ints/mouse_interfaces.cpp:433:36: note: in call to 'log10(3.200000e+01)'
        constexpr float limit_hi = std::log10(multiplicity);
                                   ^
2 warnings and 2 errors generated.
```

**Comment by FeralChild64** [comment]
OK, changed it. This code actually only runs during the startup, or if something changes mouse sensitivity, so it is not performance critical at all.

**Comment by FeralChild64** [comment]
If I understand correctly, the problem with double sensitivity was caused by the complicated mouse acceleration model present in DOSBox - it seems to me the mickey/pixel rate set by the application in the driver usually had no effect. This code still needs quite some work - for example, it updates mickey counters as soon as it receives events from the host OS, which contributes to mouse problems with Ultima Underworld.

**Comment by kcgen** [comment]
Tested in Synnergist (1996) (zip for you to try, with canned conf)

`capture_mouse = seamless` is working very nicely!

`capture_mouse = onstart middlerelease` is scaled up in the horizontal.

**Comment by FeralChild64** [comment]
Seamless used to work correctly. But I haven’t checked the Synnergist - it seems not all is fixed than.

**Comment by johnnovak** [comment]
Testing it in the evening when I get home @FeralChild64!

**Comment by johnnovak** [comment]
These are findings @FeralChild64:

Mouse movement is now perfect in:
- Eric the Unready

Doubled horizontal sensitivity is still present in:
- Discworld
- Shannara
- Knights of Xentar

So the issue is completely fixed in Eric the Unready, but unfortunately there's no improvement whatsoever in the other three. I've only tried the default PS/2 mouse.

Well, that's still progress!

**Comment by johnnovak** [comment]
Okay @FeralChild64, I think raising this bug report was a mistake on my part. I assumed the doubled mouse sensitivity issue is not present on real hardware, but now that I actually tested Eric the Unready on my new Pentium MMX box, it turns out that in fact it is, so the behaviour of Staging is correct. I tested it with a Microsoft PS/2 mouse + Cutemouse.

I will confirm later whether the issue is present in the other three games too (the answer is most likely a "yes"); for now let's just assume that there is no "fix" needed for this issue. If the difference between horizontal & vertical mouse sensitivity bothers some people (like yours truly), they can just set different horiz & vert sensitivity values in the config. But we shouldn't not mess around with the mouse emulation behaviour, because that could cause incompatibility issues.

Again, there's no way of knowing this without having access to real hardware...
